packages/events/createResponderEventPlugin.js

```diff
@@ -0,0 +1,696 @@
+ * - These events that are returned from `extractEvents` are "deferred
+ *   dispatched events".
+ * - When returned from `extractEvents`, deferred-dispatched events contain an
+ *   "accumulation" of deferred dispatches.
+ * - These deferred dispatches are accumulated/collected before they are
+ *   returned, but processed at a later time by the `EventPluginHub` (hence the
+ *   name deferred).
+ *
+ * In the process of returning their deferred-dispatched events, event plugins
+ * themselves can dispatch events on-demand without returning them from
+ * `extractEvents`. Plugins might want to do this, so that they can use event
+ * dispatching as a tool that helps them decide which events should be extracted
+ * in the first place.
+ *
+ * "On-Demand-Dispatched Events":
+ *
+ * - On-demand-dispatched events are not returned from `extractEvents`.
+ * - On-demand-dispatched events are dispatched during the process of returning
+ *   the deferred-dispatched events.
+ * - They should not have side effects.
+ * - They should be avoided, and/or eventually be replaced with another
+ *   abstraction that allows event plugins to perform multiple "rounds" of event
+ *   extraction.
+ *
+ * Therefore, the sequence of event dispatches becomes:
+ *
+ * - `R`s on-demand events (if any)   (dispatched by `R` on-demand)
+ * - `S`s on-demand events (if any)   (dispatched by `S` on-demand)
+ * - `C`s on-demand events (if any)   (dispatched by `C` on-demand)
+ * - `R`s extracted events (if any)   (dispatched by `EventPluginHub`)
+ * - `S`s extracted events (if any)   (dispatched by `EventPluginHub`)
+ * - `C`s extracted events (if any)   (dispatched by `EventPluginHub`)
+ *
+ * In the case of `ResponderEventPlugin`: If the `startShouldSetResponder`
+ * on-demand dispatch returns `true` (and some other details are satisfied) the
+ * `onResponderGrant` deferred dispatched event is returned from
+ * `extractEvents`. The sequence of dispatch executions in this case
+ * will appear as follows:
+ *
+ * - `startShouldSetResponder` (`ResponderEventPlugin` dispatches on-demand)
+ * - `touchStartCapture`       (`EventPluginHub` dispatches as usual)
+ * - `touchStart`              (`EventPluginHub` dispatches as usual)
+ * - `responderGrant/Reject`   (`EventPluginHub` dispatches as usual)
+ */
+
+type GlobalResponderHandlerType = {
+  onChange: (from: any, to: any, blockNativeResponder: any) => void,
+};
+
+type GlobalInteractionHandlerType = {
+  onChange: (numberActiveTouches: number) => void,
+};
+
+export default function createResponderEventPlugin(TopLevelTypes: {
+  topMouseDown: TopLevelType,
+  topMouseMove: TopLevelType,
+  topMouseUp: TopLevelType,
+  topScroll: TopLevelType,
+  topSelectionChange: TopLevelType,
+  topTouchCancel: TopLevelType,
+  topTouchEnd: TopLevelType,
+  topTouchMove: TopLevelType,
+  topTouchStart: TopLevelType,
+}) {
+  function isStartish(topLevelType) {
+    return (
+      topLevelType === TopLevelTypes.topMouseDown ||
```

Comment:
Please destructure these early. We don't want the overhead of object access.

Reply:
Actually I think it might make more sense to use our static forking mechanism for this